test/main.hs

```diff
@@ -78,6 +83,20 @@ main = do
       ]
     ]
 
+byteArrayCompareProp :: QC.Property
+byteArrayCompareProp = QC.property $ \(xs :: [Word8]) (ys :: [Word8]) ->
+  compareLengthFirst xs ys === compare (byteArrayFromList xs) (byteArrayFromList ys)
+
+byteArrayEqProp :: QC.Property
+byteArrayEqProp = QC.property $ \(xs :: [Word8]) (ys :: [Word8]) ->
+  (compareLengthFirst xs ys == EQ) === (byteArrayFromList xs == byteArrayFromList ys)
+
+compareLengthFirst :: [Word8] -> [Word8] -> Ordering
+compareLengthFirst xs ys = case compare (length xs) (length ys) of
+  LT -> LT
+  GT -> GT
+  EQ -> compare xs ys
```

Comment:
It might be more idiomatic to write

```haskell
compareLengthFirst xs ys = (compare `on` length) xs ys <> compare xs ys
```

Reply:
That's certainly more concise. I'll switch to that.